### DISCUSSION TITLE
How to sign ROM for secure boot ?

### DISCUSSION THREAD

**Discussion by DenisVASI9** [ipxe/ipxe#1296] [General]
Hello,

I am working with a Mellanox network card and I would like to sign its ROM with my own custom certificate to enable Secure Boot. My goal is to enroll my certificate into the motherboard's UEFI and ensure that the system can boot with the signed ROM.

However, I'm unsure about the process to sign the ROM for Mellanox hardware and integrate it into the Secure Boot workflow. Specifically:

What tools are required for signing a non-PE ROM file for Mellanox hardware?

**Comment by NiKiZe** [comment]
Converted to discussion, issues is for actual verified problems with the iPXE codebase.

There is some commits and other discussions for how to sign efi binaries, however, I have not seen anything about ROMs, also meaning, I wouldn't expect anything to really differ between efi program and efirom signing.

**Comment by mcb30** [comment]
UEFI ROMs contain a PE executable and a trivial ROM header.  Given a signed PE file, you can use iPXE's `util/efirom` to construct the ROM header.

I don't know off-hand what tools you can use to extract the PE image from an existing ROM, but you might find something in the EDK2 codebase.

**Reply by DenisVASI9** [reply]
@mcb30 Can I perform the signature at the build stage? 
Where should I add a command for signing, e.g. sbtool

**Comment by DenisVASI9** [comment]
@mcb30 I've tried doing it this way, but it has no effect
```
$(BIN)/%.efi : $(BIN)/%.efi.tmp $(ELF2EFI)
	$(QM)$(ECHO) "  [FINISH EFI] $@"
	$(Q)$(SIGNTOOL) --key $(KEY) --cert $(CERT) --output $@ $@
	$(QM)$(ECHO) "  [SIGNED] Successfully signed $@"
	$(Q)$(ELF2EFI) --subsystem=10 $< $@

$(BIN)/%.efidrv : $(BIN)/%.efidrv.tmp $(ELF2EFI)
	$(QM)$(ECHO) "  [FINISH] $@"
	$(Q)$(ELF2EFI) --subsystem=11 $< $@

$(BIN)/%.drv.efi : $(BIN)/%.efidrv
	$(QM)$(ECHO) "  [FINISH DRV] $@"
	$(Q)$(SIGNTOOL) --key $(KEY) --cert $(CERT) --output $@ $@
	$(QM)$(ECHO) "  [SIGNED] Successfully signed $@"
	$(Q)$(CP) $< $@
```

**Reply by DenisVASI9** [reply]
I mean, it seems that these lines of code are just not called, judging by the logs. Maybe I should call the signature command somewhere else.

**Comment by mcb30** [comment]
Don't try to modify the build process to accommodate `signtool`.  The build process is complex, and without a much deeper understanding of how Makefiles work you will spend hours or days on it, only to end up with something that won't ever be merged upstream since `signtool` is Windows-specific.

Instead, build the `.efidrv` variant of your build target (i.e. swap `.efirom` for `.efidrv` in your build command).  You can then use `signtool` (or equivalent tools) to sign this, and then use `./util/efirom` to build a ROM from your signed binary.
